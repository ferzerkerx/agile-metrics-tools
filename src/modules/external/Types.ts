export interface ExternalConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  srcPath?: string;
  inlineData?: ExternalData[];
}

export class ExternalData {
  createdAt: Date;
  teamName: string;
  [type: string]: any;
}

export interface ExternalService {
  fetchExternalInfo(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}

export interface ExternalRepository {
  csv(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}
