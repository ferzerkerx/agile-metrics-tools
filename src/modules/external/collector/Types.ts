import { CollectorConfig, MetricItem } from '../../../metrics';

export interface ExternalMetricItem extends MetricItem {}

export class ExternalCollectorConfig implements CollectorConfig {
  since: Date;
  until?: Date;
  teamName: string;
  type: string;
  srcType: string;
  srcPath: string;
  metricType: string;
  constructor({
    since,
    until = null,
    type,
    srcType,
    srcPath,
    metricType,
    teamName
  }) {
    this.since = since ? new Date(since) : null;
    this.until = until ? new Date(until) : null;
    this.type = type;
    this.srcType = srcType;
    this.srcPath = srcPath;
    this.metricType = metricType;
    this.teamName = teamName;
  }
}
