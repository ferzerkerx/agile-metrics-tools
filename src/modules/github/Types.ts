export class GithubCommit {
  sha: string;
  createdAt: Date;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  message: string;
  pullRequest?: PullRequest;
}

export class PullRequest {
  id: number;
  numberOfDaysOpen: number;
  numberOfComments: number;
}

export interface GithubConfig {
  repositoryName: string;
  orgName: string;
  since: string;
}

export interface GithubService {
  commits(githubConfig: GithubConfig): Promise<GithubCommit[]>;
}

export interface GithubClient {
  commits(githubConfig: GithubConfig): Promise<string[]>;

  getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<any>;

  pullRequestForCommit(sha: string): Promise<any>;
}
