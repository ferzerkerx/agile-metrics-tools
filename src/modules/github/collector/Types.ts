import { CollectorConfig, MetricItem } from '../../../metrics';

export interface GithubMetricItem extends MetricItem {
  sha: string;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  repositoryName: string;
  pullRequest?: {
    prId: number;
    numberOfDaysOpen: number;
    numberOfComments: number;
  };
}

export class GithubCollectorConfig implements CollectorConfig {
  repositoryName: string;
  orgName: string;
  since: string;

  constructor({ repositoryName, orgName, since }) {
    this.repositoryName = repositoryName;
    this.orgName = orgName;
    this.since = since;
  }
}
