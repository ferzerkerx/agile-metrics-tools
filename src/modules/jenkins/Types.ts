export class JenkinsJob {
  name: string;
  branch: string;
  builds: JenkinsBuild[];
}

export class JenkinsBuild {
  timestamp: Date;
  result: string;
  durationInMs: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  url: string;
}

export interface JenkinsService {
  findData(
    orgName: string,
    projectName: string,
    since: Date
  ): Promise<JenkinsJob>;
}

export interface JenkinsClient {
  getData(url): Promise<any>;
}
