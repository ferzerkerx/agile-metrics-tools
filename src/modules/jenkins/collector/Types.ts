import { CollectorConfig, MetricItem } from "../../../metrics/Types";

export interface JenkinsMetricItem extends MetricItem {
  jenkinsUrl: string;
  buildName: string;
  result: string;
  durationInSeconds: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  projectName: string;
}

export class JenkinsCollectorConfig implements CollectorConfig {
  orgName: string;
  projectName: string;
  since: Date;

  constructor({ orgName, projectName, since }) {
    this.orgName = orgName;
    this.projectName = projectName;
    this.since = new Date(since);
  }
}
