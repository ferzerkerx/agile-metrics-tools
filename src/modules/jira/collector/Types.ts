import { CollectorConfig, MetricItem } from '../../../metrics';

export interface JiraMetricItem extends MetricItem {
  key: string;
  createdBy: string;
  issueType: string;
  movedForward: number;
  movedBackward: number;
  storyPoints: number;
  assignees: string[];
  tags: string[];
  created: Date;
  finished: Date;
  leadTime: number;
  devTime: number;
  commentCount: number;
  jiraProject: string;
  teamName: string;
  estimateHealth: number;
  rawEstimateHealth: number;
  numberOfBugs: number;
}

export class JiraCollectorConfig implements CollectorConfig {
  teamId: number;
  teamName: string;
  workFlowType: 'sprint' | 'kanban';
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
  estimateConfig?: { maxTime: number; estimationValues: number[] };

  constructor({
    teamId,
    since,
    until = null,
    workFlowMap,
    fields,
    teamName,
    workFlowType
  }) {
    this.teamId = teamId;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.workFlowMap = workFlowMap;
    this.fields = fields;
    this.teamName = teamName;
    this.workFlowType = workFlowType ? workFlowType : 'sprint';
  }

  isKanban() {
    return this.workFlowType === 'kanban';
  }

  isSprint() {
    return this.workFlowType === 'sprint';
  }
}
