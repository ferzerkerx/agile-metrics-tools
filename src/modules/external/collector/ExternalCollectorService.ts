import { CollectorService } from '../../../metrics';
import { ExternalCollectorConfig, ExternalMetricItem } from './Types';
import { ExternalMetricConverter } from './ExternalMetricConverter';
import { ExternalConfig, ExternalService } from '../Types';

export class ExternalCollectorService
  implements CollectorService<ExternalCollectorConfig, ExternalMetricItem> {
  constructor(private externalService: ExternalService) {}

  async fetch(
    externalCollectorConfig: ExternalCollectorConfig
  ): Promise<ExternalMetricItem[]> {
    if (!this.supports(externalCollectorConfig)) {
      return [];
    }

    const externalConfig: ExternalConfig = { ...externalCollectorConfig };

    const externalData = await this.externalService.fetchExternalInfo(
      externalConfig
    );
    return ExternalMetricConverter.toMetricItem(
      externalData,
      externalCollectorConfig
    );
  }

  supports(config: any): boolean {
    return config instanceof ExternalCollectorConfig;
  }
}
