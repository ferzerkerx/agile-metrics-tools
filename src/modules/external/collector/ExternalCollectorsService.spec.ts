import { ExternalCollectorService } from './ExternalCollectorService';
import { ExternalCollectorConfig } from './Types';
import { ExternalConfig, ExternalData, ExternalService } from '../Types';

function testExternalData(): ExternalData {
  return {
    createdAt: new Date('2018-12-06'),
    field1: 'someAdditionalData',
    field2: 4,
    field3: new Date('2018-12-07')
  };
}

describe('ExternalCollectorsService', () => {
  const externalService: ExternalService = {
    fetchExternalInfo: async (
      externalConfig: ExternalConfig
    ): Promise<ExternalData[]> => {
      return [testExternalData()];
    }
  };

  const externalCollectorService: ExternalCollectorService = new ExternalCollectorService(
    externalService
  );

  it('should fetch githubMetrics', async () => {
    const externalCollectorConfig: ExternalCollectorConfig = new ExternalCollectorConfig(
      {
        since: '2018-11-20',
        type: 'someType',
        srcType: 'file',
        srcPath: 'some/path',
        metricType: 'someType'
      }
    );

    const data = await externalCollectorService.fetch(externalCollectorConfig);
    expect(data).toMatchSnapshot();
  });
});
