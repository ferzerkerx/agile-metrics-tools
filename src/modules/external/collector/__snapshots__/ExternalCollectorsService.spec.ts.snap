// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalCollectorsService should fetch githubMetrics 1`] = `
Array [
  Object {
    "createdAt": 2018-12-06T00:00:00.000Z,
    "dataType": "EXT-someType",
    "field1": "someAdditionalData",
    "field2": 4,
    "field3": 2018-12-07T00:00:00.000Z,
    "id": "ef45cd18b51f62229eea7d9e66542631a1d34893af3cc0b0616d23e0db14ae8e",
    "teamName": "someTeamName",
  },
]
`;
